dangerzone/updater/signatures.py

```diff
@@ -0,0 +1,638 @@
+
+    return True
+
+
+def get_last_log_index() -> int:
+    SIGNATURES_PATH.mkdir(parents=True, exist_ok=True)
+    if not LAST_LOG_INDEX.exists():
+        return BUNDLED_LOG_INDEX
+
+    with open(LAST_LOG_INDEX) as f:
+        return int(f.read())
+
+
+def get_log_index_from_signatures(signatures: List[Dict]) -> int:
+    def _reducer(accumulator: int, signature: Dict) -> int:
+        try:
+            logIndex = int(signature["Bundle"]["Payload"]["logIndex"])
+        except (KeyError, ValueError):
+            log.debug("Discarding invalid logindex")
+            return accumulator
+        return max(accumulator, logIndex)
+
+    return reduce(_reducer, signatures, 0)
+
+
+def write_log_index(log_index: int) -> None:
+    last_log_index_path = SIGNATURES_PATH / "last_log_index"
+
+    with open(last_log_index_path, "w") as f:
+        f.write(str(log_index))
+
+
+def _get_images_only_manifest(input: dict) -> dict:
+    """Filter out all the non-images from a loaded manifest"""
+    output = input.copy()
+    output["manifests"] = [
+        manifest
+        for manifest in input["manifests"]
+        if manifest["annotations"].get("kind")
+        in ("dev.cosignproject.cosign/imageIndex", "dev.cosignproject.cosign/image")
+    ]
+    return output
+
+
+def _get_blob(digest: str) -> PurePath:
+    return PurePath() / "blobs" / "sha256" / digest.replace("sha256:", "")
+
+
+def _get_signature_filename(input: Dict) -> PurePath:
+    for manifest in input["manifests"]:
+        if manifest["annotations"].get("kind") == "dev.cosignproject.cosign/sigs":
+            return _get_blob(manifest["digest"])
+    raise errors.SignatureExtractionError()
+
+
+def upgrade_container_image_airgapped(
+    container_tar: Path,
+    pubkey: Path = DEFAULT_PUBKEY_LOCATION,
+    bypass_logindex: bool = False,
+) -> str:
+    """
+    Verify the given archive against its self-contained signatures, then
+    upgrade the image and retag it to the expected tag.
+
+    The logic supports both "dangerzone archives" and "cosign archives".
+    The presence of a `dangerzone.json` file at the root of the tarball
+    meaning it's a "dangerzone archive".
+
+    See `prepare_airgapped_archive` for more details.
+
+    :return: The loaded image name
+    """
+
+    with TemporaryDirectory() as _tempdir, tarfile.open(container_tar, "r") as archive:
+        # First, check that we have a "signatures.json" file
+        files = archive.getnames()
+        tmppath = Path(_tempdir)
+
+        has_dangerzone_manifest = f"./{DANGERZONE_MANIFEST}" in files
+        if not has_dangerzone_manifest:
+            raise errors.InvalidImageArchive()
+
+        # Ensure that the signatures.json is the same as the index.json
+        # with only the images remaining, to avoid situations where we
+        # check the signatures but the index.json differs, making us
+        # think that we're with valid signatures where we indeed aren't.
```

Comment:
I think the `signatures.json` is a stale thing, right? Also, do we actually support both "cosign archives" and "dangerzone archives"? I'm asking because we immediately raise an exception when the archive does not have a `dangerzone.json` file (which is the correct thing to do btw).

Reply:
You're correct: `signatures.json` is now called `dangerzone.json`

We only support dangerzone archives.